components/fraud-proof-vm/fraud_proof_vm.md

```diff
@@ -0,0 +1,50 @@
+# Fraud Proof VM
+
+A fraud proof VM (FPVM) is a virtual machine which can be evaluated on-chain to settle a disputed block hash assertion. In this spec we assume the fraud proof VM is evaluated using a multi-round dispute game similar to [Truebit](https://people.cs.uchicago.edu/~teutsch/papers/truebit.pdf).
+
+## Key Properties
+
+A Fraud proof VM must satisfy the following properties to be suitable for a multi-round dispute game:
+1. Program execution can be split up into deterministic state transitions.
+2. Each state transition can be evaluated on-chain with less than or equal to a constant size gas allocation (eg. 10 million L1 gas).
+3. The FPVM must be executable off-chain in such a way that state checkpoints (ie. merklized commitments to the VM state) can be generated.
+
+## Usage
+
+### Initialization
+
+The fraud proof VM is initialized with:
+1. an executable binary; and
+2. an initialization hash for the preimage oracle.
+
+The executable binary defines the program which is executed during a fraud proof, and the initialization hash is used to load data from the FPVM's database called the `Preimage Oracle`.
+
+
+The fraud proof VM does not have access to a traditional database. Instead, it has access to the "preimage oracle". The preimage oracle is simply a database which exposes the function `get_preimage(hash: bytes)`:
+
+```python
+def get_preimage(hash: bytes) -> bytes:
+    # The preimage is looked up from a local database held by the fraud prover
+    preimage = fetch_preimage(hash)
+    assert keccak256(preimage) == hash
+    return preimage
+```
+
+The preimage oracle must be populated with all `hash->preimage` mappings before they are queried by the FPVM. These preimages are committed to in the initialization hash which is supplied as one of the two inputs to the FPVM. During the course of execution, the FPVM must **never** request a preimage which is unavailable as that would result in the FPVM being indeterminate. In this way, a valid executable binary must account for the structure of the initialization hash and only query available preimages.
```

Comment:
> must never request a preimage that is unavailable

Nevertheless, this could happen due to user actions. What happens then? --> This probably depends on the context in which the FPVM is run!

Reply:
Also, I think it's good to be a bit more precise for "are committed to in the initialization hash" (btw,extra "in". Could be good to say that all hash->preimage mapping must be proven against the initialization hash using a sequence of Merkle proofs. Probably good to link to a document that explains this, preferrably with some examples.